scio-core/src/main/scala/com/spotify/scio/transforms/ScalaAsyncLookupDoFn.scala

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2019 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.transforms
+
+import java.util.UUID
+import java.util.concurrent.{ConcurrentHashMap, ConcurrentLinkedQueue}
+
+import org.apache.beam.sdk.values.KV
+import org.apache.beam.sdk.transforms.DoFn
+import org.apache.beam.sdk.transforms.DoFn.{FinishBundle, ProcessElement, Setup, StartBundle}
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow
+import org.joda.time.Instant
+import com.google.common.base.Preconditions
+import com.spotify.scio.transforms.AsyncLookupDoFn.CacheSupplier
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Await, Future, Promise}
+import scala.concurrent.duration.Duration
+import scala.util.{Failure, Success, Try}
+import scala.util.control.NonFatal
+
+object ScalaAsyncLookupDoFn {
+
+  private val Logger = LoggerFactory.getLogger(this.getClass)
+
+  implicit private class FutureExtension[A](val future: Future[A]) {
```

Comment:
This is not resolved. Why are we converting between Scala and Java future?

Reply:
This is not converting between Scala and Jaba future, it providing the transform functionality available in Scala 2.12 since this code will also be compiled for 2.11. This is needed so transform both success and failure to Success because we are call Await.result in finishBundle which assumes that a future is successful.